### PR TITLE
Fixed #25105 -- Checked deferred constraints before updating rows on PostgreSQL.

### PR DESCRIPTION
**Pull request by David-Wobrock**
:wave: 

Ticket https://code.djangoproject.com/ticket/25105

It is possible that the change will introduce other edge cases on postgresql, for instance when the AlterField is followed by changes that would require deferred constraints.

### CODE REVIEW

File: tests/migrations/test_operations.py [hunk 1]
```diff
@@ -4058,6 +4058,68 @@ def test_alter_fk_non_fk(self):
...
+            ),
+        ]
+        new_state = self.apply_operations(app_label, project_state, operations)
+
+        Rider = new_state.apps.get_model(app_label, "Rider")
+        rider = Rider.objects.get()
+        self.assertEqual(rider.first_name, "")
+        self.assertEqual(rider.last_name, "")
+        self.assertColumnNotNull(app_label + "_rider", "first_name")
+        self.assertColumnNotNull(app_label + "_rider", "last_name")
```
Comment (felixxm) [inline tests/migrations/test_operations.py:4121]: Can we move this to the `schema` tests?, e.g.
```diff
diff --git a/tests/schema/tests.py b/tests/schema/tests.py
index 2d12796cbd..37ea0330c8 100644
--- a/tests/schema/tests.py
+++ b/tests/schema/tests.py
@@ -973,6 +973,33 @@ class SchemaTests(TransactionTestCase):
             Node.objects.update(parent=parent)
             editor.alter_field(Node, old_field, new_field, strict=True)
 
+    @isolate_apps("schema")
+    def test_alter_null_with_default_value_deferred_constraints(self):
+        class Publisher(Model):
+            class Meta:
+                app_label = "schema"
+
+        class Article(Model):
+            publisher = ForeignKey(Publisher, CASCADE)
+            title = CharField(max_length=50, null=True)
+
+            class Meta:
+                app_label = "schema"
+
+        with connection.schema_editor() as editor:
+            editor.create_model(Publisher)
+            editor.create_model(Article)
+        self.isolated_local_models = [Publisher, Article]
+
+        publisher = Publisher.objects.create()
+        Article.objects.create(publisher=publisher)
+        old_field = Article._meta.get_field("title")
+        new_field = CharField(max_length=50, null=False, default="")
+        new_field.set_attributes_from_name('title')
+        with connection.schema_editor() as editor:
+            Article.objects.update(title=None)
+            editor.alter_field(Article, old_field, new_field, strict=True)
+
     def test_alter_text_field_to_date_field(self):
         """
         #25002 - Test conversion of text field to date field.

```

**Review by felixxm — COMMENTED** [review]
@David-Wobrock Thanks :+1:

File: tests/migrations/test_operations.py (hunk 1 above)
Comment (David-Wobrock) [reply tests/migrations/test_operations.py:4121]: Sure :+1: 
I'm never quite sure yet where the most relevant place for tests is :)

Comment (felixxm) [reply tests/migrations/test_operations.py]: Fix is in `schema`, so `schema` tests seem to be a better fit.

**Comment by felixxm** [comment]
@David-Wobrock Thanks for updates :+1:

File: tests/migrations/test_operations.py (hunk 1 above)
Comment (David-Wobrock) [reply tests/migrations/test_operations.py]: Of course, makes sense :sweat_smile: 
Thanks for the review and merge :ship:
